File: cli/npm/src/commands/registry.ts
import { Command } from 'commander';
import chalk from 'chalk';
import * as fs from 'fs-extra';
import * as path from 'path';
import { RegistryClient } from '../lib/registry';
import { ConfigManager } from '../lib/config';

export function createRegistryCommand(): Command {
  const command = new Command('registry');
  command.description('Manage package registries');

  // Registry list command
  const listCommand = new Command('list');
  listCommand
    .description('List all configured registries')
    .action(async () => {
      try {
        const configManager = ConfigManager.getInstance();
        const config = await configManager.loadConfig();
        const registries = config.registry.registries;
        const defaultRegistry = config.registry.default;

        if (Object.keys(registries).length === 0) {
          console.log(chalk.yellow('No registries configured'));
          console.log(chalk.dim("Use 'prompd registry add' to add a registry"));
          return;
        }

        console.log(chalk.bold('Configured Registries:'));
        for (const [name, registryConfig] of Object.entries(registries)) {
          const defaultMark = name === defaultRegistry ? chalk.green(' (default)') : '';
          const url = registryConfig.url || 'N/A';
          const username = registryConfig.username || 'Not logged in';
          console.log(`  • ${chalk.cyan(name)}${defaultMark}`);
          console.log(`    URL: ${url}`);
          console.log(`    User: ${username}`);
        }
      } catch (error) {
        console.error(chalk.red(`Error listing registries: ${error}`));
        process.exit(1);
      }
    });

  // Registry add command
  const addCommand = new Command('add');
  addCommand
    .description('Add a new registry')
    .argument('<name>', 'Registry name')
    .argument('<url>', 'Registry URL')
    .action(async (name: string, url: string) => {
      try {
        const configManager = ConfigManager.getInstance();
        await configManager.addRegistry(name, url);
        console.log(chalk.green(`Added registry '${chalk.cyan(name)}' at ${url}`));
      } catch (error) {
        console.error(chalk.red(`Error adding registry: ${error}`));
        process.exit(1);
      }
    });

  // Registry remove command
  const removeCommand = new Command('remove');
  removeCommand
    .description('Remove a registry')
    .argument('<name>', 'Registry name')
    .option('--force', 'Force removal of default prompdhub registry')
    .action(async (name: string, options) => {
      try {
        // Warn about removing prompdhub unless forced
        if (name === 'prompdhub' && !options.force) {
          console.log(chalk.yellow('Warning:') + ' Removing the main prompdhub registry');
          console.log('Use --force to confirm, or consider setting a different default instead');
          process.exit(1);
        }

        const configManager = ConfigManager.getInstance();
        await configManager.removeRegistry(name);
        console.log(chalk.green(`Removed registry '${chalk.cyan(name)}'`));
      } catch (error) {
        console.error(chalk.red(`Error removing registry: ${error}`));
        process.exit(1);
      }
    });

  // Registry set-default command
  const setDefaultCommand = new Command('set-default');
  setDefaultCommand
    .description('Set the default registry')
    .argument('<name>', 'Registry name')
    .action(async (name: string) => {
      try {
        const configManager = ConfigManager.getInstance();
        await configManager.setDefaultRegistry(name);
        console.log(chalk.green(`Set '${chalk.cyan(name)}' as default registry`));
      } catch (error) {
        console.error(chalk.red(`Error setting default registry: ${error}`));
        process.exit(1);
      }
    });

  // Add all registry management commands
  command.addCommand(listCommand);
  command.addCommand(addCommand);
  command.addCommand(removeCommand);
  command.addCommand(setDefaultCommand);

  return command;
}

// Legacy publish command (to be integrated with main CLI)
export function createPublishCommand(): Command {
  const publishCommand = new Command('publish');
  publishCommand
    .description('Publish a package to the registry')
    .argument('[directory]', 'Directory containing project.prmdproj file', '.')
    .option('--access <access>', 'Package access level', 'public')
    .option('--tag <tag>', 'Package tag', 'latest')
    .option('--dry-run', 'Show what would be published without actually publishing')
    .option('--force', 'Force publish even if version exists')
    .option('--registry <registry>', 'Registry name to publish to')
    .action(async (directory: string, options) => {
      try {
        console.log(chalk.cyan(`📦 Publishing package from ${directory}...`));
        
        // Resolve registry
        const registryName = options.registry;
        const client = new RegistryClient(registryName);

        // Event listeners for progress
        client.on('publishStart', ({ packageDir, options }) => {
          console.log(chalk.blue(`🚀 Starting publish from ${packageDir}`));
        });
        
        client.on('publishComplete', ({ name, version, access }) => {
          console.log(chalk.green(`✅ Successfully published ${name}@${version} (${access})`));
          console.log(chalk.gray(`   Registry: ${client.registryUrl}`));
          console.log(chalk.gray(`   Install: prompd install ${name}@${version}`));
        });
        
        client.on('publishError', ({ packageDir, error }) => {
          console.error(chalk.red(`❌ Publish failed: ${error.message}`));
        });

        await client.publish(path.resolve(directory), {
          access: options.access as 'public' | 'private',
          tag: options.tag,
          dryRun: options.dryRun,
          force: options.force
        });

      } catch (error) {
        console.error(chalk.red(`Publish failed: ${error}`));
        process.exit(1);
      }
    });

  return publishCommand;
}

// Legacy search command (to be integrated with main CLI)
export function createSearchCommand(): Command {
  const searchCommand = new Command('search');
  searchCommand
    .description('Search for packages in the registry')
    .argument('<query>', 'Search query')
    .option('-l, --limit <limit>', 'Maximum number of results', '20')
    .option('-r, --registry <registry>', 'Search specific registry')
    .action(async (query: string, options) => {
      try {
        const limit = parseInt(options.limit, 10);
        const registryName = options.registry;
        const client = new RegistryClient(registryName);

        const results = await client.search({ query, limit });

        if (results.packages.length === 0) {
          console.log(`No packages found for: ${chalk.cyan(query)}`);
          if (registryName) {
            console.log(chalk.dim(`Searched in registry: ${registryName}`));
          }
          return;
        }

        const registryDisplayName = registryName || 'default registry';
        console.log(`Found ${results.packages.length} package(s) for ${chalk.cyan(query)} in ${chalk.yellow(registryDisplayName)}:`);
        console.log();

        for (const pkg of results.packages) {
          console.log(`${chalk.bold.cyan(pkg.name)} - ${chalk.green(`v${pkg.version}`)}`);
          console.log(`  ${pkg.description || 'No description available'}`);
          
          if (pkg.author) {
            console.log(`  Author: ${pkg.author}`);
          }
          
          console.log();
        }

      } catch (error) {
        console.error(chalk.red(`Search failed: ${error}`));
        process.exit(1);
      }
    });

  return searchCommand;
}

// Legacy install command (to be integrated with main CLI)
export function createInstallCommand(): Command {
  const installCommand = new Command('install');
  installCommand
    .description('Install a package from the registry')
    .argument('<package>', 'Package name')
    .option('--version <version>', 'Specific version to install')
    .option('--global', 'Install globally')
    .option('--registry <registry>', 'Install from specific registry (overrides scope resolution)')
    .action(async (packageName: string, options) => {
      try {
        const configManager = ConfigManager.getInstance();
        
        // Resolve which registry to use
        let registryName: string;
        if (options.registry) {
          // Explicit registry override
          registryName = options.registry;
        } else {
          // Auto-resolve based on package scope
          registryName = configManager.resolveRegistryForPackage(packageName);
        }

        const client = new RegistryClient(registryName);

        console.log(`Installing package: ${chalk.cyan(packageName)}`);
        console.log(`   Registry: ${chalk.yellow(registryName)}`);
        if (options.version) {
          console.log(`   Version: ${chalk.green(options.version)}`);
        }

        await client.install(packageName, {
          version: options.version,
          global: options.global
        });

        console.log(chalk.green('Package installed successfully!'));

      } catch (error) {
        console.error(chalk.red(`Installation failed: ${error}`));
        process.exit(1);
      }
    });

  return installCommand;
}

// Login command (to be integrated with main CLI)
export function createLoginCommand(): Command {
  const loginCommand = new Command('login');
  loginCommand
    .description('Authenticate with a Prompd registry')
    .option('-t, --token <token>', 'Use API token instead of interactive login')
    .option('-r, --registry <registry>', 'Login to specific registry')
    .action(async (options) => {
      try {
        const registryName = options.registry;
        const client = new RegistryClient(registryName);
        const actualRegistryName = registryName || 'default';

        if (options.token) {
          // Token-based login
          console.log(chalk.dim(`Authenticating with ${actualRegistryName}...`));
          const userData = await client.loginWithToken(options.token);
          
          console.log(chalk.green(`Successfully authenticated as ${userData.username}`));
          console.log(`   Registry: ${actualRegistryName} (${client.registryUrl})`);
        } else {
          // Interactive login not implemented yet
          console.log(chalk.yellow('Interactive login not yet implemented. Use --token option.'));
          process.exit(1);
        }

      } catch (error) {
        console.error(chalk.red(`Authentication failed: ${error}`));
        process.exit(1);
      }
    });

  return loginCommand;
}

// Logout command (to be integrated with main CLI)
export function createLogoutCommand(): Command {
  const logoutCommand = new Command('logout');
  logoutCommand
    .description('Clear registry authentication credentials')
    .option('-r, --registry <registry>', 'Logout from specific registry')
    .action(async (options) => {
      try {
        const registryName = options.registry;
        const client = new RegistryClient(registryName);
        
        await client.logout();
        
        const actualRegistryName = registryName || 'default';
        console.log(chalk.green(`Successfully logged out from ${actualRegistryName}`));

      } catch (error) {
        console.error(chalk.red(`Logout failed: ${error}`));
        process.exit(1);
      }
    });

  return logoutCommand;
}

File: go/cmd/prompd/registry.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

// PackageInfo represents package information from registry
type PackageInfo struct {
	Name           string                   `json:"name"`
	Version        string                   `json:"version"`
	LatestVersion  string                   `json:"latest_version"`
	Description    string                   `json:"description"`
	Author         string                   `json:"author"`
	License        string                   `json:"license"`
	Tags           []string                 `json:"tags"`
	Parameters     []map[string]interface{} `json:"parameters"`
	CreatedAt      string                   `json:"created_at"`
	UpdatedAt      string                   `json:"updated_at"`
	DownloadCount  int                      `json:"download_count"`
}

func handleRegistry() {
	if len(os.Args) < 3 {
		printRegistryUsage()
		return
	}

	switch os.Args[2] {
	case "list":
		handleRegistryList()
	case "add":
		handleRegistryAdd()
	case "remove":
		handleRegistryRemove()
	case "set-default":
		handleRegistrySetDefault()
	case "login":
		handleRegistryLogin()
	case "logout":
		handleRegistryLogout()
	case "search":
		handleRegistrySearch()
	case "info":
		handleRegistryInfo()
	case "publish":
		handleRegistryPublish()
	case "install":
		handleRegistryInstall()
	case "versions":
		handleRegistryVersions()
	default:
		fmt.Printf("Unknown registry command: %s\n", os.Args[2])
		printRegistryUsage()
	}
}

func printRegistryUsage() {
	fmt.Println("Usage: prompd registry <command> [options]")
	fmt.Println("")
	fmt.Println("Registry management commands:")
	fmt.Println("  list                     List all configured registries")
	fmt.Println("  add <name> <url>         Add a new registry")
	fmt.Println("  remove <name> [--force]  Remove a registry")
	fmt.Println("  set-default <name>       Set the default registry")
	fmt.Println("")
	fmt.Println("Registry operations:")
	fmt.Println("  login [-r <registry>] -t <token>  Authenticate with registry")
	fmt.Println("  logout [-r <registry>]            Clear authentication")
	fmt.Println("  search <query> [-r <registry>]    Search packages")
	fmt.Println("  info <package> [-r <registry>]    Get package information")
	fmt.Println("  publish [--registry <name>]       Publish package")
	fmt.Println("  install <package> [-r <registry>] Install package")
	fmt.Println("  versions <package> [-r <registry>] List package versions")
}

// Registry Management Commands

func handleRegistryList() {
	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	registries := getRegistriesMap(config)
	defaultRegistry := ""
	if def, exists := config.Registry["default"]; exists {
		if defStr, ok := def.(string); ok {
			defaultRegistry = defStr
		}
	}

	if len(registries) == 0 {
		fmt.Println("No registries configured")
		fmt.Println("Use 'prompd registry add' to add a registry")
		return
	}

	fmt.Println("Configured Registries:")
	for name, regData := range registries {
		defaultMark := ""
		if name == defaultRegistry {
			defaultMark = " (default)"
		}

		url := "N/A"
		username := "Not logged in"

		if regMap, ok := regData.(map[string]interface{}); ok {
			if urlVal, exists := regMap["url"]; exists {
				if urlStr, ok := urlVal.(string); ok {
					url = urlStr
				}
			}
			if usernameVal, exists := regMap["username"]; exists {
				if usernameStr, ok := usernameVal.(string); ok {
					username = usernameStr
				}
			}
		}

		fmt.Printf("  • %s%s\n", name, defaultMark)
		fmt.Printf("    URL: %s\n", url)
		fmt.Printf("    User: %s\n", username)
	}
}

func handleRegistryAdd() {
	if len(os.Args) < 5 {
		fmt.Println("Usage: prompd registry add <name> <url>")
		os.Exit(1)
	}

	name := os.Args[3]
	url := os.Args[4]

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	if err := addRegistry(config, name, url); err != nil {
		fmt.Printf("Error adding registry: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Added registry '%s' at %s\n", name, url)
}

func handleRegistryRemove() {
	if len(os.Args) < 4 {
		fmt.Println("Usage: prompd registry remove <name> [--force]")
		os.Exit(1)
	}

	name := os.Args[3]
	force := false

	// Check for --force flag
	for _, arg := range os.Args[4:] {
		if arg == "--force" {
			force = true
			break
		}
	}

	// Warn about removing prompdhub unless forced
	if name == "prompdhub" && !force {
		fmt.Println("Warning: Removing the main prompdhub registry")
		fmt.Println("Use --force to confirm, or consider setting a different default instead")
		os.Exit(1)
	}

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	if err := removeRegistry(config, name); err != nil {
		fmt.Printf("Error removing registry: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Removed registry '%s'\n", name)
}

func handleRegistrySetDefault() {
	if len(os.Args) < 4 {
		fmt.Println("Usage: prompd registry set-default <name>")
		os.Exit(1)
	}

	name := os.Args[3]

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	if err := setDefaultRegistry(config, name); err != nil {
		fmt.Printf("Error setting default registry: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Set '%s' as default registry\n", name)
}

// Authentication Commands

func handleRegistryLogin() {
	var registryName string
	var token string

	// Parse arguments
	for i := 3; i < len(os.Args); i++ {
		switch os.Args[i] {
		case "-r", "--registry":
			if i+1 < len(os.Args) {
				registryName = os.Args[i+1]
				i++
			}
		case "-t", "--token":
			if i+1 < len(os.Args) {
				token = os.Args[i+1]
				i++
			}
		}
	}

	if token == "" {
		fmt.Println("Usage: prompd registry login [-r <registry>] -t <token>")
		os.Exit(1)
	}

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	// Resolve registry name
	if registryName == "" {
		if def, exists := config.Registry["default"]; exists {
			if defStr, ok := def.(string); ok {
				registryName = defStr
			}
		}
		if registryName == "" {
			registryName = "prompdhub"
		}
	}

	// Get registry info
	regInfo, err := getRegistryInfo(config, registryName)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Authenticating with %s...\n", registryName)

	// Verify token by getting user profile
	resp, err := http.Get(fmt.Sprintf("%s/v1/user/me", regInfo.URL))
	if err != nil {
		fmt.Printf("Authentication failed: %v\n", err)
		os.Exit(1)
	}
	defer resp.Body.Close()

	req, err := http.NewRequest("GET", fmt.Sprintf("%s/v1/user/me", regInfo.URL), nil)
	if err != nil {
		fmt.Printf("Authentication failed: %v\n", err)
		os.Exit(1)
	}
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))

	client := &http.Client{}
	resp, err = client.Do(req)
	if err != nil {
		fmt.Printf("Authentication failed: %v\n", err)
		os.Exit(1)
	}
	defer resp.Body.Close()

	if resp.StatusCode == 401 {
		fmt.Println("Invalid token. Please check your API token.")
		os.Exit(1)
	} else if resp.StatusCode == 404 {
		fmt.Printf("User endpoint not found. Registry URL: %s\n", regInfo.URL)
		os.Exit(1)
	} else if resp.StatusCode >= 500 {
		fmt.Printf("Registry server error (%d). Please try again later.\n", resp.StatusCode)
		os.Exit(1)
	}

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		fmt.Printf("Authentication failed: %s\n", string(body))
		os.Exit(1)
	}

	var userData map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&userData); err != nil {
		fmt.Printf("Invalid response from registry: %v\n", err)
		os.Exit(1)
	}

	username, ok := userData["username"].(string)
	if !ok {
		fmt.Println("Invalid response from registry: missing username")
		os.Exit(1)
	}

	// Save token
	if err := setRegistryToken(config, registryName, token, username); err != nil {
		fmt.Printf("Error saving token: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Successfully authenticated as %s\n", username)
	fmt.Printf("   Registry: %s (%s)\n", registryName, regInfo.URL)
}

func handleRegistryLogout() {
	var registryName string

	// Parse arguments
	for i := 3; i < len(os.Args); i++ {
		switch os.Args[i] {
		case "-r", "--registry":
			if i+1 < len(os.Args) {
				registryName = os.Args[i+1]
				i++
			}
		}
	}

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	// Resolve registry name
	if registryName == "" {
		if def, exists := config.Registry["default"]; exists {
			if defStr, ok := def.(string); ok {
				registryName = defStr
			}
		}
		if registryName == "" {
			registryName = "prompdhub"
		}
	}

	// Clear token
	if err := setRegistryToken(config, registryName, "", ""); err != nil {
		fmt.Printf("Error clearing token: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Successfully logged out from %s\n", registryName)
}

// Registry Operation Commands

func handleRegistrySearch() {
	if len(os.Args) < 4 {
		fmt.Println("Usage: prompd registry search <query> [-r <registry>]")
		os.Exit(1)
	}

	query := os.Args[3]
	var registryName string

	// Parse optional registry argument
	for i := 4; i < len(os.Args); i++ {
		switch os.Args[i] {
		case "-r", "--registry":
			if i+1 < len(os.Args) {
				registryName = os.Args[i+1]
				i++
			}
		}
	}

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	// Resolve registry name
	if registryName == "" {
		if def, exists := config.Registry["default"]; exists {
			if defStr, ok := def.(string); ok {
				registryName = defStr
			}
		}
		if registryName == "" {
			registryName = "prompdhub"
		}
	}

	// Get registry info
	regInfo, err := getRegistryInfo(config, registryName)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}

	// Perform search
	url := fmt.Sprintf("%s/v1/packages?q=%s", regInfo.URL, query)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		fmt.Printf("Search failed: %v\n", err)
		os.Exit(1)
	}

	if regInfo.Token != "" {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", regInfo.Token))
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("Search failed: %v\n", err)
		os.Exit(1)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		fmt.Printf("Search failed (%d): %s\n", resp.StatusCode, string(body))
		os.Exit(1)
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		fmt.Printf("Invalid response: %v\n", err)
		os.Exit(1)
	}

	packages, ok := result["packages"].([]interface{})
	if !ok {
		fmt.Printf("Invalid response format\n")
		os.Exit(1)
	}

	if len(packages) == 0 {
		fmt.Printf("No packages found for: %s\n", query)
		if registryName != "prompdhub" {
			fmt.Printf("Searched in registry: %s\n", registryName)
		}
		return
	}

	fmt.Printf("Found %d package(s) for '%s' in %s:\n\n", len(packages), query, registryName)

	for _, pkgData := range packages {
		if pkgMap, ok := pkgData.(map[string]interface{}); ok {
			name := getString(pkgMap, "name", "Unknown")
			version := getString(pkgMap, "version", "0.0.0")
			description := getString(pkgMap, "description", "No description available")
			author := getString(pkgMap, "author", "")

			fmt.Printf("%s - v%s\n", name, version)
			fmt.Printf("  %s\n", description)
			if author != "" {
				fmt.Printf("  Author: %s\n", author)
			}
			fmt.Println()
		}
	}
}

func handleRegistryInfo() {
	if len(os.Args) < 4 {
		fmt.Println("Usage: prompd registry info <package> [-r <registry>]")
		os.Exit(1)
	}

	packageName := os.Args[3]
	var registryName string

	// Parse optional registry argument
	for i := 4; i < len(os.Args); i++ {
		switch os.Args[i] {
		case "-r", "--registry":
			if i+1 < len(os.Args) {
				registryName = os.Args[i+1]
				i++
			}
		}
	}

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	// Auto-resolve registry if not specified
	if registryName == "" {
		registryName = resolveRegistryForPackage(config, packageName)
	}

	// Get registry info
	regInfo, err := getRegistryInfo(config, registryName)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}

	// Get package info
	url := fmt.Sprintf("%s/v1/packages/%s", regInfo.URL, packageName)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		fmt.Printf("Request failed: %v\n", err)
		os.Exit(1)
	}

	if regInfo.Token != "" {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", regInfo.Token))
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("Request failed: %v\n", err)
		os.Exit(1)
	}
	defer resp.Body.Close()

	if resp.StatusCode == 404 {
		fmt.Printf("Package not found: %s\n", packageName)
		os.Exit(1)
	}

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		fmt.Printf("Request failed (%d): %s\n", resp.StatusCode, string(body))
		os.Exit(1)
	}

	var pkgInfo map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&pkgInfo); err != nil {
		fmt.Printf("Invalid response: %v\n", err)
		os.Exit(1)
	}

	// Display package information
	name := getString(pkgInfo, "name", "Unknown")
	version := getString(pkgInfo, "version", "0.0.0")
	description := getString(pkgInfo, "description", "No description")
	author := getString(pkgInfo, "author", "Unknown")
	license := getString(pkgInfo, "license", "Unknown")

	fmt.Printf("Package: %s v%s\n", name, version)
	fmt.Printf("Description: %s\n", description)
	fmt.Printf("Author: %s\n", author)
	fmt.Printf("License: %s\n", license)
	fmt.Printf("Registry: %s\n", registryName)
}

func handleRegistryPublish() {
	var registryName string

	// Parse optional registry argument
	for i := 3; i < len(os.Args); i++ {
		switch os.Args[i] {
		case "--registry":
			if i+1 < len(os.Args) {
				registryName = os.Args[i+1]
				i++
			}
		}
	}

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	// Resolve registry name
	if registryName == "" {
		if def, exists := config.Registry["default"]; exists {
			if defStr, ok := def.(string); ok {
				registryName = defStr
			}
		}
		if registryName == "" {
			registryName = "prompdhub"
		}
	}

	// Get registry info
	regInfo, err := getRegistryInfo(config, registryName)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}

	if regInfo.Token == "" {
		fmt.Printf("Not authenticated with registry '%s'. Use 'prompd registry login -r %s -t <token>' first.\n", registryName, registryName)
		os.Exit(1)
	}

	fmt.Printf("Publishing to registry: %s (%s)\n", registryName, regInfo.URL)
	fmt.Println("Note: Publish implementation requires package files - placeholder for now")
}

func handleRegistryInstall() {
	if len(os.Args) < 4 {
		fmt.Println("Usage: prompd registry install <package> [-r <registry>]")
		os.Exit(1)
	}

	packageName := os.Args[3]
	var registryName string

	// Parse optional registry argument
	for i := 4; i < len(os.Args); i++ {
		switch os.Args[i] {
		case "-r", "--registry":
			if i+1 < len(os.Args) {
				registryName = os.Args[i+1]
				i++
			}
		}
	}

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	// Auto-resolve registry if not specified
	if registryName == "" {
		registryName = resolveRegistryForPackage(config, packageName)
	}

	// Get registry info
	regInfo, err := getRegistryInfo(config, registryName)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Installing package: %s\n", packageName)
	fmt.Printf("   Registry: %s (%s)\n", registryName, regInfo.URL)
	fmt.Println("Note: Install implementation requires download logic - placeholder for now")
}

func handleRegistryVersions() {
	if len(os.Args) < 4 {
		fmt.Println("Usage: prompd registry versions <package> [-r <registry>]")
		os.Exit(1)
	}

	packageName := os.Args[3]
	var registryName string

	// Parse optional registry argument
	for i := 4; i < len(os.Args); i++ {
		switch os.Args[i] {
		case "-r", "--registry":
			if i+1 < len(os.Args) {
				registryName = os.Args[i+1]
				i++
			}
		}
	}

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	// Auto-resolve registry if not specified
	if registryName == "" {
		registryName = resolveRegistryForPackage(config, packageName)
	}

	// Get registry info
	regInfo, err := getRegistryInfo(config, registryName)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Versions for package: %s\n", packageName)
	fmt.Printf("   Registry: %s (%s)\n", registryName, regInfo.URL)
	fmt.Println("Note: Versions implementation requires API call - placeholder for now")
}

// Helper function to safely get string from map
func getString(m map[string]interface{}, key, defaultVal string) string {
	if val, exists := m[key]; exists {
		if str, ok := val.(string); ok {
			return str
		}
	}
	return defaultVal
}

// getPackageInfo fetches package information from registry
func getPackageInfo(packageName string, registryName string) (*PackageInfo, error) {
	config, err := LoadConfig()
	if err != nil {
		return nil, fmt.Errorf("failed to load config: %w", err)
	}

	regInfo, err := getRegistryInfo(config, registryName)
	if err != nil {
		return nil, err
	}

	// Construct API URL
	url := fmt.Sprintf("%s/packages/%s", regInfo.URL, packageName)

	// Make HTTP request
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Add authentication if available
	if regInfo.Token != "" {
		req.Header.Set("Authorization", "Bearer "+regInfo.Token)
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch package info: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("registry returned status %d: %s", resp.StatusCode, string(body))
	}

	var info PackageInfo
	if err := json.NewDecoder(resp.Body).Decode(&info); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &info, nil
}

// getPackageVersions fetches available versions for a package
func getPackageVersions(packageName string, registryName string) ([]string, error) {
	config, err := LoadConfig()
	if err != nil {
		return nil, fmt.Errorf("failed to load config: %w", err)
	}

	regInfo, err := getRegistryInfo(config, registryName)
	if err != nil {
		return nil, err
	}

	// Construct API URL
	url := fmt.Sprintf("%s/packages/%s/versions", regInfo.URL, packageName)

	// Make HTTP request
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Add authentication if available
	if regInfo.Token != "" {
		req.Header.Set("Authorization", "Bearer "+regInfo.Token)
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch versions: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("registry returned status %d: %s", resp.StatusCode, string(body))
	}

	var versions []string
	if err := json.NewDecoder(resp.Body).Decode(&versions); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return versions, nil
}